python/adjoint/objective.py

```diff
@@ -76,28 +79,39 @@ def place_adjoint_source(self,dJ,dt):
             dV = 1/self.sim.resolution * 1/self.sim.resolution
         else:
             dV = 1/self.sim.resolution * 1/self.sim.resolution * 1/self.sim.resolution
-        da_dE = 0.5*(dV * self.cscale)
-        scale = da_dE * dJ * 1j * 2 * np.pi * self.frequencies / np.array([self.time_src.fourier_transform(f) for f in self.frequencies]) # final scale factor
+        
+        da_dE = 0.5 * self.cscale # scalar popping out of derivative
+        iomega = (1.0 - np.exp(-1j * (2 * np.pi * self.frequencies) * dt)) * (1.0 / dt) # scaled frequency factor with discrete time derivative fix
+
         if self.frequencies.size == 1:
             # Single frequency simulations. We need to drive it with a time profile.
             src = self.time_src
-            amp = scale
+            
+            # an ugly way to calcuate the scaled dtft of the forward source
+            y = np.array([src.swigobj.current(t,dt) for t in np.arange(0,T,dt)]) # time domain signal
+            fwd_dtft = np.matmul(np.exp(1j*2*np.pi*self.frequencies[:,np.newaxis]*np.arange(y.size)*dt), y)*dt/np.sqrt(2*np.pi) # dtft
+            
+            # we need to compensate for the phase added by the time envelope at our freq of interest
+            src_center_dtft = np.matmul(np.exp(1j*2*np.pi*np.array([src.frequency])[:,np.newaxis]*np.arange(y.size)*dt), y)*dt/np.sqrt(2*np.pi)
+            adj_src_phase = np.exp(1j*np.angle(src_center_dtft))
+            
+            amp = da_dE * dV * dJ * iomega / fwd_dtft / adj_src_phase # final scale factor
         else:
-            # TODO: In theory we should be able drive the source without normalizing out the time profile.
-            # But for some reason, there is a frequency dependent scaling discrepency. It works now for 
-            # multiple monitors and multiple sources, but we should figure out why this is.
-            src = FilteredSource(self.time_src.frequency,self.frequencies,scale,dt,self.time_src) # generate source from broadband response
+            # multi frequency simulations
+            scale = da_dE * dV * dJ * iomega # final scale factor
```

Comment:
There's still a bug with this **broadband** scale factor (the single frequency case is fine). The amplitude is correct, but there is still a phase error. The resulting gradients are "less correlated" than they should be. The gradient accuracy actually *decreases* with resolution.

Here's a sample gradient distribution (adjoint vs finite difference) when resolution = 10
![image](https://user-images.githubusercontent.com/24902086/88586550-8e1dbb00-d022-11ea-8d11-65fa7c33fe39.png)


and when resolution = 30
![image](https://user-images.githubusercontent.com/24902086/88586865-08e6d600-d023-11ea-9661-3990a0f8f7fd.png)

I'm assuming I need to normalize the phase of the input Gaussian profile (i.e. divide by `fwd_dtft` like in the single frequency case) -- unfortunately, that only makes it worse.

Any thoughts?

Reply:
Ah, the issue appears to either be with how the forward/adjoint fields are stored, or how the gradient is calculated (`meepgeom.py`). The gradient corresponding to the first frequency in the array of gradients is always correct, regardless of which frequency it is or what the resolution is. I/ve tried randomly choosing different frequencies, shuffling their order, etc. -- the first frequency is always correct and the others diverge as resolution increases.

I've tested the `FilteredSource` routine and that is consistently correct. My guess would be there is something wrong with how I'm iterating through the array of gradients in C++ before returning them back to python...

Reply:
Aha!

Commenting out the patch submitted in #1233 fixes the issues.

Reply:
Per our discussion, this should be resolved by #1306 — which is now merged so you should rebase this PR onto master.